src/main.ts

```diff
@@ -0,0 +1,250 @@
+ * MetaMask repositories will be (or have been) cloned.
+ * @param args.config.defaultProjectNames - The set of MetaMask repositories
+ * that will be linted.
+ */
+export async function main({
+  argv,
+  stdout,
+  stderr,
+  config: { cachedRepositoriesDirectoryPath, defaultProjectNames },
+}: {
+  argv: string[];
+  stdout: SimpleWriteStream;
+  stderr: SimpleWriteStream;
+  config: {
+    cachedRepositoriesDirectoryPath: string;
+    defaultProjectNames: string[];
+  };
+}) {
+  const outputLogger = new OutputLogger({ stdout, stderr });
+  const workingDirectoryPath = WORKING_DIRECTORY_PATH;
+
+  const inputs = await parseInputs({ argv, outputLogger, defaultProjectNames });
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
+  if (!inputs) {
+    process.exitCode = 1;
+    return;
+  }
+  const { templateRepositoryName, projectReferences } = inputs;
+
+  const template = await establishMetaMaskRepository({
+    repositoryReference: templateRepositoryName,
+    workingDirectoryPath,
+    cachedRepositoriesDirectoryPath,
+    outputLogger,
+  });
+
+  await lintProjects({
+    projectReferences,
+    template,
+    workingDirectoryPath,
+    cachedRepositoriesDirectoryPath,
+    outputLogger,
+  });
+}
+
+/**
+ * Parses command-line arguments and establishes parameters for the remainder of
+ * this tool.
+ *
+ * @param args - The arguments to this function.
+ * @param args.argv - The name of this executable and its arguments (as obtained
+ * via `process.argv`).
+ * @param args.outputLogger - Writable streams for output messages.
+ * @param args.defaultProjectNames - The set of MetaMask repositories
+ * that will be linted.
+ * @returns The parsed command-line arguments if parsing was successful, or null
+ * otherwise.
+ */
+async function parseInputs({
+  argv,
+  outputLogger,
+  defaultProjectNames,
+}: {
+  argv: string[];
+  outputLogger: OutputLogger;
+  defaultProjectNames: string[];
+}): Promise<Inputs | null> {
+  const args = await parseCommandLineArguments({ argv, outputLogger });
+
+  if (args) {
+    const projectReferences = args._.length > 0 ? args._ : defaultProjectNames;
+    return {
+      templateRepositoryName: DEFAULT_TEMPLATE_REPOSITORY_NAME,
+      projectReferences,
+    };
+  }
+
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
+  return null;
+}
+
+/**
+ * Parses the command line using `yargs`.
+ *
+ * @param args - The arguments to this function.
+ * @param args.argv - The name of this executable and its arguments (as obtained
+ * via `process.argv`).
+ * @param args.outputLogger - Writable streams for output messages.
+ */
+async function parseCommandLineArguments({
+  argv,
+  outputLogger,
+}: {
+  argv: string[];
+  outputLogger: OutputLogger;
+}) {
+  let yargsFailure: { message: string; error: Error } | null = null;
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
```

Comment:
If failure is not possible, why has this been written to anticipate it?

Also, it seems that strict is enabled (I think it's on by default), so any unrecognized flag/command would be treated as an error. So it should be possible to encounter this.

Reply:
Ah, I see the line below on the `_` option. This seems to suggest that it is accepting unrecognized arguments, contrary to what the docs suggest. I mist be missing something here.

Reply:
I realized I hadn't actually tested failure cases. I added one [here]([3c5ec57](https://github.com/MetaMask/module-lint/pull/33/commits/3c5ec5715e2fa52b8251015030bd9fa1c3d64d49)), and that meant I could remove one of the ignore lines. I kept this one, but moved it down. I'm not sure how Yargs generates errors, but I figured that if `onFail` takes an `error` argument we should do _something_ exceptional with it. (I could have used it further down in the `if (yargsFailure)` block, but for some reason TypeScript wants to type `yargsFailure` as `never`. I'm not quite sure why at the moment. I'll try playing around with this further.)

Reply:
I had missed that strict mode is disabled by default, OK. Using strict options makes sense though, good idea